File: book/petrol-pump/chapter3/section3/AccumulatePulsesPump.java
package chapter3.section3;

import pump.*;
import chapter3.section2.LifeCycle;
import sodium.*;
import java.util.Optional;

public class AccumulatePulsesPump implements Pump {
    public Outputs create(Inputs inputs) {
        LifeCycle lc = new LifeCycle(inputs.eNozzle1,
                                     inputs.eNozzle2,
                                     inputs.eNozzle3);
        Behavior<Double> litersDelivered =
                accumulate(lc.eStart.map(u -> Unit.UNIT),
                           inputs.eFuelPulses,
                           inputs.calibration);
        return new Outputs()
            .setDelivery(lc.fillActive.map(
                of ->
                    of.equals(Optional.of(Fuel.ONE))   ? Delivery.FAST1 :
                    of.equals(Optional.of(Fuel.TWO))   ? Delivery.FAST2 :
                    of.equals(Optional.of(Fuel.THREE)) ? Delivery.FAST3 : 
                                                         Delivery.OFF))
            .setSaleQuantityLCD(litersDelivered.map(
                    q -> Formatters.formatSaleQuantity(q)));
    }

    public static Behavior<Double> accumulate(
            Event<Unit> eClearAccumulator,
            Event<Integer> ePulses,
            Behavior<Double> calibration) {
        BehaviorLoop<Integer> total = new BehaviorLoop<>();
        total.loop(ePulses.snapshot(total,
            (pulses_, total_) -> pulses_ + total_)
               .merge(eClearAccumulator.map(f -> 0))
               .hold(0));
        return Behavior.lift(
            (total_, calibration_) -> total_ * calibration_,
            total, calibration);
    }
}

File: c#/Book/PetrolPump/PetrolPump/PetrolPump/Chapter3/Section5/AccumulatePulsesPump.cs
using System;

using PetrolPump.Chapter3.Section3;
using PetrolPump.Pump;

using Sodium;

namespace PetrolPump.Chapter3.Section5
{
public class AccumulatePulsesPump : IPump 
{
    public Outputs Create(Inputs inputs) 
    {
        LifeCycle lc = new LifeCycle(inputs.ENozzle1,
                                     inputs.ENozzle2,
                                     inputs.ENozzle3);
        Behavior<Double> litersDelivered =
                Accumulate(lc.EStart.Map(u => Unit.UNIT),
                           inputs.EFuelPulses,
                           inputs.Calibration);
        return new Outputs()
            .SetDelivery(lc.FillActive.Map(
                Of =>
                    Of.Equals(Optional<Fuel>.Of(Fuel.ONE))   ? Delivery.FAST1 :
                    Of.Equals(Optional<Fuel>.Of(Fuel.TWO))   ? Delivery.FAST2 :
                    Of.Equals(Optional<Fuel>.Of(Fuel.THREE)) ? Delivery.FAST3 : 
                                                         Delivery.OFF))
            .SetSaleQuantityLcd(litersDelivered.Map(
                    q => Formatters.FormatSaleQuantity(q)));
    }

    public static Behavior<double> Accumulate(
            Event<Unit> eClearAccumulator,
            Event<int> ePulses,
            Behavior<double> calibration) {
        BehaviorLoop<int> total = new BehaviorLoop<int>();
        total.Loop(ePulses.Snapshot(total,
            (pulses_, total_) => pulses_ + total_)
               .Merge(eClearAccumulator.Map(f => 0))
               .Hold(0));
        return Behavior<double>.Lift(
            (total_, calibration_) => total_ * calibration_,
            total, calibration);
    }
  }
}
